### ISSUE TITLE
Fix `declaration-property-max-values` false positives for interpolated inline expressions

### ISSUE THREAD

**Issue by mindplay-dk** [stylelint/stylelint#9422]
### What minimal example or steps are needed to reproduce the bug?

```jsx
import styled from '@emotion/styled'

export const TableCell = styled(({ noPadding, ...rest }) => <div {...rest} />)`
  padding: ${({ noPadding }) => (noPadding ? '0' : '1em')};
`
```

### What minimal configuration is needed to reproduce the bug?

```mjs
/** @type {import("stylelint").Config} */
export default {
  customSyntax: 'postcss-styled-syntax',
  rules: {
    'declaration-property-max-values': {
      '/^margin$/': 2,
      '/^padding$/': 2,
      '/^border-width$/': 2,
      '/^border-style$/': 2,
      '/^border-color$/': 2,
    },
  },
}
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICWBbADhATgFwAQGdsBPAGxilwDNMIVcByAARhQmyQgDsB6Q08+gB1OwmAA8MOXJE6FcAFQCGAIzIBhGCRK4AvAWJkoACiPBcnCAAVFUKEk4BzADS4AdO8ww5AXwCUugD5cAB47ADdcYHdXT0JvXG4A3wADYVxcO3w0EkUieCoyMQBuNNxFEiQHTgBaJGwWfHywGE56zBLOdLQbO0d8gBJgU3MrHvsHXD9A3CMLa1tx3AB+BgAGelx8+gBGFnpfbw7UzhAnEEokMgA5RRQ4RABlAy9XACt8MVPwLguHBBBuAAqQG4RjENAwSKoSTYIyCEB8MgVVrw3yuNQ-SrxQHcUQSLB4WCURQAVxIeGApTAJMItAeRFaijEWwwhDA+Hw1UR5C5DOwTPoTlKmDJXnylM66QYsDAOUwinYXGqaBoEJwRGqKCZ1TC5RJXno4tKUoY3AAelrMA57P1uIbcAAmIWSk30c3dBaOW32p3GqVus3KLCwTDVADuSCg2AAFt78r6Xf7zUHMCGuc8447nSb0gGU2nICQsJmEybvNnJs7vMIvjJfgAxLBa7D-bnI7CuOuVVwod5fWBofD-CXpeFt+zYeH5eE5eqEeEVsfPdvVLsOdOKThQRSpqe4GcKryT06leGs7DszncqC8xliPcHufH4TVk7eIA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "latest",
  "postcss-styled-syntax": "latest"
}
```

### What did you expect to happen?

Both branches are valid - no problems should be reported.

### What actually happened?

Apparently, any CSS-in-JS expression triggers this error, for example:

```js
import styled from '@emotion/styled'

export const TableCell = styled(({ noPadding, ...rest }) => <div {...rest} />)`
  padding: ${() => '0'};
`
```

### Do you have a proposal to fix the bug?

Ignore expressions that aren't supported, can't be parsed or evaluated.

**Comment by mindplay-dk** [comment]
Here's a more minimal repro: [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICWBbADhATgFwAQGdsBPAGxilwDNMIVcByAARhQmyQgDsB6Q08+gB1OqDDlzBcYfPlwBfKjTpMWbDj0wwAhmGxDOwmAA8xeSJ0K4AKloBGZAMIwSJXAF4CxMlAB0UJABuAAbCuLhoWlD+nADm8LgAJMAAFACU7gB8DACMLPRyANzCIZwgADQglEhkAHJaKHCIAMpeMPg+AFb4RuXgXFUxCCDcAFQjuIzEaDASoljYyYIgfGQkSJzYS6k+Dv1IMQoj3IYm87iwlFoAriR4wKFSV4S0TUQbWkbx9BiE0vgAtCtyIC3tgPvQyg9MDc2vF7pwwmF6LAwCQtJgtOwuP80DRpjgiP8UB9-gEtCQrm16HCHoiGBEousYtTcABmSEIxFyDlhbnCOTCXrmAYAMSwxOwQyBaw2PmF+x8KC6vVgaHwQ3hYSW0vWmwQuCWaOwbT1PINIB+2D+gNaUBB7yMS3ihsxJqW-MFciAA)

If you change the `padding` value in `declaration-property-max-values` in config from `3` to `4`, it clears.

So even with a single expression wrapped in a function, for some reason it thinks there are `3` values. 🤔

**Comment by mindplay-dk** [comment]
Another minimal example to narrow it down: [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICWBbADhATgFwAQGdsBPAGxilwDNMIVcByAARhQmyQgDsB6Q08+gB1OqDDlzBcYfPlwBfKjTpMWbDj0wwAhmGxDOwyJ0IFaMAMIQSWXAF4GAYi0AjMLEr7hMAB5i8RkwAVFzJzGBISOwJiMigAOigkADcAA2FcXGcsWEwAWkhrTHhcABJgfDNLQtwAagYAFgAGejkAbmE0zhAAGhBKJDIAOS0UOEQAZRiYfDiAK3xvHvAufoBzBBBuACot3EZiNBgJUSxsAApBED4yEiRObEuASjjLTjWFLe4vX1Pcdy0AK4kPDAdJSAGEWjjIj3LTeYr0DCEaT4XLXchomHYOH0bpgzBA6bFUGcDIZeiwMAkLSYLTsLi5NA0Q44Ii5FBw3JJLQkAHTejEsFkhhZTA5fJWLAC3AAJjxpLJcnlGSVwjkwiWRjWADEsBzsBt0bd7nEtUhVnEUPMlrA0PgNiSMpcjXcHghcJdqdhpm7lR6QEjsCi0VMoJjYd5LsVPXSfZc1Rq5EA)

Change the `someColor + '40'` expression to just `someColor`, and it clears.

The limit in this example is `2` - so it appears to treat `someColor + '40'` as two values.

It appears to treat any literal value as a CSS value - as though it doesn't really parse the expressions at all.

**Comment by mindplay-dk** [comment]
I think I've isolated the problem - it looks like a bug in `postcss-styled-syntax`, and I may have a working fix.

I will open a PR to discuss with the maintainer of that - I will link it here and post an update if they confirm the issue.

**Comment by jeddy3** [comment]
@mindplay-dk Thank you for your report and for investigating. Yes, it does seem to be a bug in the `postcss-styled-syntax` syntax. Thanks for opening a PR there. Let's see what comes of it.

**Comment by mindplay-dk** [comment]
@jeddy3 we have an [answer](https://github.com/hudochenkov/postcss-styled-syntax/pull/39#issuecomment-5228615358) from @hudochenkov 

> The current behavior was specifically designed this way. That's the whole point of a custom syntax. There are many cases where we need to check the value exactly as it is present in the source file.

I guess that makes sense? 🤔

> The approach the Stylelint team has taken over the years is to exclude non-standard syntax from the rule check when there is a chance it could cause false positives. The best solution, in my opinion, is to add [`isStandardSyntaxValue`](https://github.com/stylelint/stylelint/blob/main/lib/utils/isStandardSyntaxValue.mjs) check to the [`declaration-property-max-values`](https://github.com/stylelint/stylelint/blob/main/lib/rules/declaration-property-max-values/index.mjs) rule code.

I don't guess there's any hope of being able to actually tokenize JS at these sites? Probably way too complex, right? 😊

So the question is what's worse, I guess?

Do we prefer a ton of false positives we have to check and manually suppress?

Or do we prefer a lot of potentially real problems flying under the radar with no encouragement to check?

I think, if we were to make this rule simply ignore JS interpolations, that would be considered a breaking change, right?

And either way, it should probably be user's choice, which trade off they prefer? e.g. an option to ignore non-CSS syntax, but switched off by default, so it's not a breaking change.

I'm honestly not sure if I'd turn it on or not myself. Neither of these trade offs fully solve my problem. Who knew it would be this difficult to migrate to logical properties? 😅 I'll need to think about it...

**Comment by jeddy3** [comment]
@mindplay-dk Thanks for the update. 

Aleks is correct, and I was wrong about it being a bug in the custom syntax. 

As he said, we favour false negatives over positives and have applied this consistently to rules.

The built-in rules are geared towards standard CSS. People are free to create their own [custom versions](https://stylelint.io/developer-guide/plugins/) of built-in rules, e.g. as [the SCSS community does](https://github.com/stylelint-scss/stylelint-scss), to handle the complexities of non-standard syntax.

Let's use the [`isStandardSyntaxValue`](https://github.com/stylelint/stylelint/blob/main/lib/utils/isStandardSyntaxValue.mjs) utility. 

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/contributor-guide/rules#fix-a-bug-in-a-rule) in the Contributor guide.

It'll be a non-breaking and patch change, as [the rule will report fewer errors](https://stylelint.io/about/semantic-versioning).

> Who knew it would be this difficult to migrate to logical properties?

Yes, unfortunately so. I suspect the [current lack of 1:1 mappings for shorthand properties](https://drafts.csswg.org/css-logical-1/#logical-shorthand-keyword) complicated things, especially for tool authors.

**Comment by mindplay-dk** [comment]
I'm not sure I understand, what is a false negative?

By false positives, I meant it's reporting errors where there aren't any. 

As explained, I don't think think I'd want this changed, even if I could - because a lot of those reported errors *are* errors, even if some of them are not. 

Equating "we don't know" with "there's no error" would be a problem - I turned on the rule to flag any *potential* errors for review.

Not all of them lead to a fix - some lead to a refactor or a stylelint-disable directive, but then at least we know they've been reviewed.

**Comment by ybiquitous** [comment]
I agree with @jeddy3 here.

The current parser (`postcss-value-parser`) doesn't support non-standard CSS syntax, such as CSS-in-JS. See the reproduction:

```js
// repro-value-parser.mjs
import valueParser from 'postcss-value-parser';
import { isValueFunction, isValueString, isValueWord } from './lib/utils/typeGuards.mjs';

// Same predicate as lib/rules/declaration-property-max-values/index.mjs
const isValueNode = (node) => isValueWord(node) || isValueFunction(node) || isValueString(node);

const values = [
  '1em 2em',
  '${someColor}',
  "${someColor + '40'}",
  "${someColor+'40'}",
  "${() => '0'}",
  "${({ noPadding }) => (noPadding ? '0' : '1em')}",
];

for (const value of values) {
  const nodes = valueParser(value).nodes;
  const count = nodes.filter(isValueNode).length;
  console.log(`${JSON.stringify(value)} => ${count}`);
}
```

Run it on the Stylelint repo root:

```sh-session
$ node repro-value-parser.mjs
"1em 2em" => 2
"${someColor}" => 1
"${someColor + '40'}" => 4
"${someColor+'40'}" => 3
"${() => '0'}" => 4
"${({ noPadding }) => (noPadding ? '0' : '1em')}" => 4
```

Note that the parser counts 4 values in `${someColor + '40'}`, while it counts 3 values in `${someColor+'40'}`. This doesn't seem to satisfy your needs.

So I believe parsing non-standard syntax is a plugin's responsibility, not built-in rules.
